### PR TITLE
Implement pager support wrt #1118

### PR DESCRIPTION
**Pull request by hasufell**
TODO:

- [x] document
- [x] config options should be `pager`, not `pager-config` (adjust json)
- [x] deal with windows

### CODE REVIEW

File: app/ghcup/Main.hs [hunk 1]
```diff
@@ -68,17 +69,18 @@ import qualified GHCup.Types                   as Types
 toSettings :: Options -> IO (Settings, KeyBindings, UserSettings)
 toSettings options = do
   noColor <- isJust <$> lookupEnv "NO_COLOR"
+  pagerCmd <- getPager
```
Comment (hasufell) [inline app/ghcup/Main.hs:72]: We're unconditionally executing this, which is a bit unfortunate... I think lazy IO could be fine here.

**Comment by hasufell** [comment]
When using `less`... the unicode stuff doesn't work very well. You have to either export `LESSCHARSET=utf-8` or use `PAGER='less -r'`.

File: lib/GHCup/Utils/Pager.hs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        Just r  -> pure $ Just r
+        Nothing -> lookupEnv "PAGER" >>= \case
+          Just r' -> pure $ Just r'
+          Nothing -> do
+                findExecutable "most"
+            <|> findExecutable "more"
+            <|> findExecutable "less"
+
+sendToPager :: FilePath -> Text -> IO ()
+sendToPager pager text = sendToPagerConduit pager (sourceLbs (BSL.fromStrict . E.encodeUtf8 $ text))
```
Comment (hasufell) [inline lib/GHCup/Utils/Pager.hs:37]: Assuming UTF-8 here is a bit unfortunate. But I don't know if ghcup behaves well wrt unicode symbols if you have esoteric encodings set anyway.

File: lib/GHCup/Utils/Pager.hs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+sendToPagerConduit :: FilePath -> ConduitT () ByteString (ResourceT IO) () -> IO ()
+sendToPagerConduit pager producer = do
+     ((Just stdinH),_,(Just stderrH),ph) <-
+       createProcess
```
Comment (dfordivam) [inline lib/GHCup/Utils/Pager.hs:42]: I am not confident about this, for a start using `withCreateProcess` might be better. And should the std_out be `Inherit`, and delegate_ctlc = True?

I had looked at the implementation done by git and it is doing signal handling changes, like ignoring signals which ought to be handled by pager.

**Comment by dfordivam** [comment]
For windows the `more` might work, will need testing i it work from within the process's invocation, writing to stdin (https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/more)

**Comment by hasufell** [comment]
> For windows the `more` might work, will need testing i it work from within the process's invocation, writing to stdin (https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/more)

in powershell, it does not... in msys2 anything works

**Comment by dfordivam** [comment]
> When using `less`... the unicode stuff doesn't work very well. You have to either export `LESSCHARSET=utf-8` or use `PAGER='less -r'`.

I had `LESS=-R` in my config and hence didn't notice this issue.

**Comment by dfordivam** [comment]
LGTM, the incremental improvements could be done later.

**Review by dfordivam — APPROVED** [review]

File: lib/GHCup/Utils/Pager.hs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    $ \(Just stdinH) _ (Just stderrH) ph -> do
+        forM_ text $ T.hPutStrLn stdinH
+        hClose stdinH
+        exitCode <- waitForProcess ph
+        case exitCode of
+          ExitFailure i ->
+            do errContents <- hGetContents stderrH
+               fail (unlines [mappend "Pager exited with exit code " (show i)
+                             ,errContents])
+          _ -> pure ()
```
Comment (dfordivam) [inline lib/GHCup/Utils/Pager.hs:51]: I looked into the `more` support, by looking at its source code. It reads from the STDERR and thus gets stuck if we create a new pipe for std_err. I confirmed that this change fixes `more` on both linux and windows
 https://github.com/util-linux/util-linux/blob/81ea117e0396274808a77d40bf11af44ea047285/text-utils/more.c#L804C11-L804C24

```suggestion
-- 'more' reads from STDERR, and requires std_err to be 'Inherit'
sendToPager :: FilePath -> [Text] -> IO (Either IOException ())
sendToPager pager text = try @IOException
    $ withCreateProcess (shell pager) { std_in = CreatePipe
                                      , std_err = Inherit
                                      , std_out = Inherit
                                      , delegate_ctlc = True
                                      }
    $ \(Just stdinH) _ _ ph -> do
        forM_ text $ T.hPutStrLn stdinH
        hClose stdinH
        exitCode <- waitForProcess ph
        case exitCode of
          ExitFailure i -> fail ("Pager exited with exit code " <> show i)
          _ -> pure ()
```

Comment (hasufell) [reply lib/GHCup/Utils/Pager.hs:51]: Great find

**Pull request merged**